### PR TITLE
[Custom] Implement simple application  @open sesame 04/06 08:14

### LINKED ISSUE #1772: Expose custom optimizer to the public.
docs and example how to use custom optimizer to the public

### PR DESCRIPTION
**Pull request by zhoonit**
## Commits to be reviewed in this PR

<details><summary>[custom] expose optimizer example</summary><br />

This patch exposes optimizer example. This patch is just copy & paste
for now. The upcoming patch will update this to the updated format

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [X]Skipped
2. Run test: [ ]Passed [ ]Failed [X]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

</details>

<details><summary>[Custom] Update api of momentum optimizer</summary><br />

This patch update api of momentum optimizer.

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [x]Skipped
2. Run test: [ ]Passed [ ]Failed [x]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

</details>

<details><summary>[Custom] Add example skeleton for the custom opt</summary><br />

This patch add example skeleton for the custom optimizer.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

</details>

<details><summary>[Custom] Implement simple application</summary><br />

Add simple application which uses custom optimizer

resolves #1772

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

</details>

### CODE REVIEW

File: Applications/Custom/momentum.cpp [hunk 1]
```diff
@@ -6,55 +6,48 @@
...
-  int status = ML_ERROR_NONE;
-  if (key == "momentum") {
-    status = setDouble(momentum, value);
-  } else {
-    OptimizerImpl::setProperty(key, value);
-  }
+std::vector<ml::train::TensorDim>
+Momentum::getOptimizerVariableDim(const ml::train::TensorDim &dim) {
+  /// momentum optimizer uses *accumulated
+  return {dim};
```
Comment (kparichay) [inline Applications/Custom/momentum.cpp:41]: This variable is allocated for each weight with ITERATION_LIFESPAN.
How about give a lifespan option (differentiate between local and iteration memory)?

Recommend to do in a separate PR.

File: Applications/Custom/OptimizerClient/main.cpp [hunk 2]
```diff
@@ -29,5 +68,71 @@ int main() {
...
+
+  try {
+    dataset = ml::train::createDataset(ml::train::DatasetType::GENERATOR,
+                                       constant_generator_cb);
+  } catch (...) {
+    std::cerr << "creating dataset failed";
+    return 1;
+  }
+
+  if (model->setDataset(ml::train::DatasetModeType::MODE_TRAIN, dataset) != 0) {
```
Comment (kparichay) [inline Applications/Custom/OptimizerClient/main.cpp:86]: this can also throw?

File: Applications/Custom/OptimizerClient/main.cpp [hunk 3]
```diff
@@ -29,5 +68,71 @@ int main() {
...
+
+  try {
+    ///////// attaching custom momentum
+    optimizer = ml::train::createOptimizer("custom_momentum");
+  } catch (...) {
+    std::cerr << "creating optimizer failed";
+    return 1;
+  }
+
+  if (model->setOptimizer(optimizer) != 0) {
```
Comment (kparichay) [inline Applications/Custom/OptimizerClient/main.cpp:99]: same here.

File: Applications/Custom/OptimizerClient/main.cpp [hunk 4]
```diff
@@ -29,5 +68,71 @@ int main() {
...
 
+  auto model =
+    ml::train::createModel(ml::train::ModelType::NEURAL_NET, {"batch_size=10"});
+
+  std::shared_ptr<ml::train::Dataset> dataset;
+  std::shared_ptr<ml::train::Optimizer> optimizer;
+  std::vector<std::shared_ptr<ml::train::Layer>> layers;
+
+  try {
+    dataset = ml::train::createDataset(ml::train::DatasetType::GENERATOR,
```
Comment (kparichay) [inline Applications/Custom/OptimizerClient/main.cpp:79]: how about using our random data generator as the dataset here?

File: Applications/Custom/OptimizerClient/main.cpp [hunk 5]
```diff
@@ -29,5 +68,71 @@ int main() {
     return 1;
   }
 
+  auto model =
+    ml::train::createModel(ml::train::ModelType::NEURAL_NET, {"batch_size=10"});
```
Comment (kparichay) [inline Applications/Custom/OptimizerClient/main.cpp:72]: 10 -> BATCH_SIZE

File: Applications/Custom/OptimizerClient/main.cpp [hunk 6]
```diff
@@ -29,5 +68,71 @@ int main() {
...
+
+  if (model->setOptimizer(optimizer) != 0) {
+    std::cerr << "failed to set optimizer";
+    return 1;
+  }
+
+  try {
+    /// creating a simple model
+    layers = std::vector<std::shared_ptr<ml::train::Layer>>{
+      ml::train::layer::Input({"name=inputlayer", "input_shape=1:1:100"}),
```
Comment (kparichay) [inline Applications/Custom/OptimizerClient/main.cpp:107]: 100 -> FEATURE_SIZE

File: Applications/Custom/OptimizerClient/main.cpp [hunk 7]
```diff
@@ -29,5 +68,71 @@ int main() {
...
+      ml::train::layer::FullyConnected(
+        {"name=outputlayer", "input_layers=inputlayer", "unit=10",
```
Comment (kparichay) [inline Applications/Custom/OptimizerClient/main.cpp:109]: 10 -> NUM_CLASS

File: Applications/Custom/momentum.cpp [hunk 8]
```diff
@@ -6,55 +6,48 @@
...
 
-namespace nntrainer {
-
-const std::string Momentum::type = "momentum";
+namespace custom {
 
-void Momentum::addOptimizerVariable(std::vector<Weight> &weight_list) {
-  for (auto &w : weight_list) {
-    w.clearOptimizerVariables();
+Momentum::Momentum() { setProperty({"learning_rate=0.001", "momentum=0.9f"}); }
```
Comment (kparichay) [inline Applications/Custom/momentum.cpp:20]: i recommend to match the default learning rate of SGD and momentum optimizer.

File: Applications/Custom/momentum.cpp (hunk 1 above)
Comment (zhoonit) [reply Applications/Custom/momentum.cpp:41]: Got it, just confirming, isn't this max lifespan == iteration lifespan (as of now?)

File: Applications/Custom/OptimizerClient/main.cpp (hunk 2 above)
Comment (zhoonit) [reply Applications/Custom/OptimizerClient/main.cpp:86]: We should make this void to avoid double error checking :p, updated

File: Applications/Custom/OptimizerClient/main.cpp (hunk 3 above)
Comment (zhoonit) [reply Applications/Custom/OptimizerClient/main.cpp:99]: forwarding to https://github.com/nnstreamer/nntrainer/pull/1781#discussion_r768252170

File: Applications/Custom/OptimizerClient/main.cpp (hunk 4 above)
Comment (zhoonit) [reply Applications/Custom/OptimizerClient/main.cpp:79]: This is not opened for the API, for now, let's stick with it. It will be a good first issue.

File: Applications/Custom/OptimizerClient/main.cpp (hunk 5 above)
Comment (zhoonit) [reply Applications/Custom/OptimizerClient/main.cpp:72]: Done

File: Applications/Custom/OptimizerClient/main.cpp (hunk 6 above)
Comment (zhoonit) [reply Applications/Custom/OptimizerClient/main.cpp:107]: Done

File: Applications/Custom/momentum.cpp (hunk 1 above)
Comment (kparichay) [reply Applications/Custom/momentum.cpp:41]: no , iteration lifespan != max lifespan.

**Review by kparichay — APPROVED** [review]
LGTM

**Review by DonghakPark — APPROVED** [review]
LGTM

**Review by songgot — APPROVED** [review]
LGTM
